lib/source-map/mapping-list.js

```diff
@@ -0,0 +1,88 @@
+/* -*- Mode: js; js-indent-level: 2; -*- */
+/*
+ * Copyright 2014 Mozilla Foundation and contributors
+ * Licensed under the New BSD license. See LICENSE or:
+ * http://opensource.org/licenses/BSD-3-Clause
+ */
+if (typeof define !== 'function') {
+    var define = require('amdefine')(module, require);
+}
+define(function (require, exports, module) {
+
+  var util = require('./util');
+
+  /**
+   * Determine whether mappingB is after mappingA with respect to generated position.
+   */
+  function generatedPositionAfter(mappingA, mappingB) {
+    // Optimized for most common case
+    var lineA = mappingA.generatedLine;
+    var lineB = mappingB.generatedLine;
+    var columnA = mappingA.generatedColumn;
+    var columnB = mappingB.generatedColumn;
+    return lineB > lineA || lineB == lineA && columnB >= columnA ||
+           util.compareByGeneratedPositions(mappingA, mappingB) <= 0;
+  }
+
+  /**
+   * A data structure to provide a sorted view of accumulated mappings in a
+   * performance conscious manner. It trades a neglibable overhead in general
+   * case for a large speedup in case of mappings being added in order.
+   */
+  function MappingList() {
+    this._array = [];
+    this._unordered = false;
+    // Serves as infimum
+    this._last = {generatedLine: -1, generatedColumn: 0};
+  }
+
+  /**
+   * Static method for creating MappingList instances from an existing array.
+   */
+  MappingList.fromArray = function MappingList_fromArray(aArray) {
+    var list = new MappingList();
+    list._array = aArray.slice().sort(util.compareByGeneratedPositions);
+    return list;
+  };
+
+
+  /**
+   * Iterate through internal items.
+   */
+  MappingList.prototype.forEach = function MappingList_forEach() {
+    return Array.prototype.forEach.apply(this._array, arguments);
+  };
```

Comment:
How about we just rename it `unsortedForEach` or `noSortForEach` so that it doesn't imply any sorting order? Should add a blurb in the doc comment about how it doesn't ensure sorting.